ouroboros-consensus-cardano-tools/app/immdb-server.hs

```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE ApplicativeDo   #-}
+{-# LANGUAGE NamedFieldPuns  #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main (main) where
+
+import           Data.Void
+import qualified Network.Socket as Socket
+import           Options.Applicative
+
+import           Ouroboros.Consensus.Node.ProtocolInfo (ProtocolInfo (..))
+
+import           Cardano.Tools.DBAnalyser.Block.Cardano (Args (..))
+import           Cardano.Tools.DBAnalyser.HasAnalysis (mkProtocolInfo)
+import qualified Cardano.Tools.ImmDBServer.Diffusion as ImmDBServer
+
+main :: IO ()
+main = do
+  Opts {..} <- execParser optsParser
+  let sockAddr = Socket.SockAddrInet port hostAddr
+        where
+          -- could also be passed in
+          hostAddr = Socket.tupleToHostAddress (127, 0, 0, 1)
+      args = CardanoBlockArgs configFile Nothing
+  ProtocolInfo{pInfoConfig} <- mkProtocolInfo args
+  absurd <$> ImmDBServer.run immDBDir sockAddr pInfoConfig
+
+data Opts = Opts {
+    immDBDir   :: FilePath
+  , port       :: Socket.PortNumber
+  , configFile :: FilePath
+  }
+
+optsParser :: ParserInfo Opts
+optsParser =
+    info (helper <*> parse) $ fullDesc <> progDesc desc
+  where
+    desc = "Serve an ImmutableDB via ChainSync and BlockFetch"
+
+    parse = do
+      immDBDir <- strOption $ mconcat
```

Comment:
Question: minor styling question: Do we prefer monadic-style  or applicative-style for constructing parsers? eg. 
alternative would be
```
Opts 
  <$> parseDbDir
  <*> parsePort
  <*> parseConfig
```
? Something to write down somewhere as a coding convention?

Reply:
I have a weak preference for [`ApplicativeDo`+`RecordWildCards`](https://github.com/pcapriotti/optparse-applicative/#applicative-do) here, as one does not have to worry about field ordering and can also introduce local bindings; but happy to change of we want decisions like these to be made globally.